cnap/core/frame.py

```diff
@@ -0,0 +1,250 @@
+        self._ts_new = time.time()
+        self._ts_infer_start = 0
+
+    @property
+    def pipeline_id(self) -> str:
+        """
+        Get the pipeline ID.
+        """
+        return self._pipeline_id
+
+    @pipeline_id.setter
+    def pipeline_id(self, new_str: str) -> None:
+        """
+        Set pipeline ID from string
+        """
+        self._pipeline_id = new_str
+
+    @property
+    def sequence(self) -> int:
+        """
+        Get the monolithic counter for the frame sequence number.
+        """
+        return self._sequence
+
+    @property
+    def timestamp_new_frame(self) -> float:
+        """
+        Get the timestamp of the beginning of this new frame.
+        """
+        return self._ts_new
+
+    @timestamp_new_frame.setter
+    def timestamp_new_frame(self, timestamp: float) -> None:
+        """
+        Set the timestamp of the beginning of this new frame.
+        """
+        self._ts_new = timestamp
+
+    @property
+    def raw(self) -> np.ndarray:
+        """
+        Get the raw frame.
+        """
+        return self._raw
+
+    @raw.setter
+    def raw(self, raw_frame: np.ndarray) -> None:
+        """
+        Set raw frame.
+        """
+        self._raw = raw_frame
+
+    @property
+    def timestamp_infer_start(self) -> float:
+        """
+        Get the timestamp of the inference's start point.
+        """
+        return self._ts_infer_start
+
+    @timestamp_infer_start.setter
+    def timestamp_infer_start(self, timestamp: float) -> None:
+        """
+        Set the timestamp of the inference's start point.
+        """
+        self._ts_infer_start = timestamp
+
+    def to_blob(self) -> Any:
+        """
+        Encode the raw frame to a blob for transferring to the infer queue.
+
+        Returns:
+            Any: the blob encoded from raw frame.
+        """
+        # Get the frame shape info
+        if len(self._raw.shape) == 3:
+            height, width, channel = self._raw.shape
+        else:
+            height, width, channel = 1, len(self._raw), 1
+
+        frame_msg = frame_pb2.FrameMessage()
+        frame_msg.pipeline_id = self._pipeline_id
+        frame_msg.sequence = self._sequence
+        frame_msg.raw = self._raw.tobytes()
+        frame_msg.ts_new = self._ts_new
+        frame_msg.raw_height = height
+        frame_msg.raw_width = width
+        frame_msg.raw_channels = channel
+        frame_msg.ts_infer_start = self._ts_infer_start
+
+        serialized_data = frame_msg.SerializeToString()
+        # TODO: zlib, gzip or QATZip to accelerate the compression
+        compressed_data = serialized_data
+
+        return compressed_data
+
+    @staticmethod
+    def from_blob(blob: Any) -> Any:
+        """
+        Restore a frame instance from a blob.
+
+        Returns:
+            Any: the frame restored from a blob.
+        """
+        # TODO: zlib, gzip or QATZip to accelerate the decompression
+        decompressed_data = blob
```

Comment:
For protobuf serializing, we may add `try catch` to raise error.

Reply:
Thanks, fixed.